contracts/token/ERC20/extensions/draft-ERC20Votes.sol

```diff
@@ -0,0 +1,166 @@
+
+        Checkpoint[] storage ckpts = _checkpoints[account];
+
+        // Property: low and high converge on the first (earliest) checkpoint that is AFTER (or equal) `blockNumber`.
+        // - If all checkpoints are before `blockNumber`, low and high converge toward `ckpts.length`
+        // - If all checkpoints are after `blockNumber`, low and high will converge toward `0`
+        // - If there are no checkpoints, low = high = 0 = ckpts.length, and the 2 properties above do hold
+        // At each iteration:
+        // - If checkpoints[mid].fromBlock is equal or after `blockNumber`, we look in [low, mid] (mid is a candidate)
+        // - If checkpoints[mid].fromBlock is before `blockNumber`, we look in [mid+1, high] (mid is not a candidate)
+        // Once we have found the first checkpoint AFTER (or equal) `blockNumber`, we get the value at the beginning of
+        // `blockNumber` by reading the checkpoint just before that. If there is no checkpoint before, then we return 0
+        // (no value).
+        uint256 high = ckpts.length;
+        uint256 low = 0;
+        while (low < high) {
+            uint256 mid = Math.average(low, high);
+            if (ckpts[mid].fromBlock > blockNumber) {
+                high = mid;
+            } else {
+                low = mid + 1;
+            }
+        }
+
+        return high == 0 ? 0 : ckpts[high - 1].votes;
+    }
+
+    /**
+     * @dev Delegate votes from the sender to `delegatee`.
+     */
+    function delegate(address delegatee) public virtual override {
+        return _delegate(_msgSender(), delegatee);
+    }
+
+    /**
+     * @dev Delegates votes from signatory to `delegatee`
+     */
+    function delegateBySig(address delegatee, uint256 nonce, uint256 expiry, uint8 v, bytes32 r, bytes32 s)
+    public virtual override
+    {
+        require(block.timestamp <= expiry, "ERC20Votes::delegateBySig: signature expired");
+        address signatory = ECDSA.recover(
+            _hashTypedDataV4(keccak256(abi.encode(
+                _DELEGATION_TYPEHASH,
+                delegatee,
+                nonce,
+                expiry
+            ))),
+            v, r, s
+        );
+        require(nonce == _useNonce(signatory), "ERC20Votes::delegateBySig: invalid nonce");
+        return _delegate(signatory, delegatee);
+    }
+
+    /**
+     * @dev Change delegation for `delegator` to `delegatee`.
+     */
+    function _delegate(address delegator, address delegatee) internal virtual {
+        address currentDelegate = delegates(delegator);
+        uint256 delegatorBalance = balanceOf(delegator);
+        _delegates[delegator] = delegatee;
+
+        emit DelegateChanged(delegator, currentDelegate, delegatee);
+
+        _moveDelegates(currentDelegate, delegatee, delegatorBalance);
+    }
+
+    function _moveDelegates(address srcRep, address dstRep, uint256 amount) private {
+        if (srcRep != dstRep && amount > 0) {
```

Comment:
I think the `Rep` variable names are confusing. The "rep" concept isn't used anywhere else in the code. 

Also `_moveDelegates` may be better named `_moveVotingPower`?

Reply:
Names come from Compound. I agree they are not great, if you have better ideas I'm all hears